ENH: release 1.12 branch (#323)
data-apis/array-api-compat: docs/changelog.md

# Changelog

## 1.12.0 (2025-05-13)

### Major changes

- The build system has been updated to use `pyproject.toml` instead of `setup.py`
- Support for Python 3.9 has been dropped. The minimum supported Python version is now

>   3.10; the minimum supported NumPy version is 1.22; the minimum supported `ndonnx`
>   version is 0.10.1.

Reviewer: ```suggestion
  3.10; the minimum supported NumPy version is 1.22.
```
I find this special-casing for ndonnx a bit bizarre. We're not testing against anything other than the latest versions.
As a general rule, I would not advertise what I do not test.

> - The `linalg` extension works correctly with `pytorch==2.7`.

Reviewer: ```suggestion
- The `linalg` extension works correctly with `pytorch>=2.7`.
```

> - Multiple improvements to handling of `device` arguments in `numpy`, `cupy`, `torch`,
>   and `dask` backends. Support for multiple devices is still relatively immature,
>   and rough edges can be expected. Please report any issues you encounter.

Reviewer: ```suggestion
- Multiple improvements to handling of devices in CuPy and PyTorch backends.
  Support for multiple devices in CuPy is still immature and you should use
  context managers rather than relying on input-output device propagation or
  on the `device` parameter.
```
We didn't do anything in numpy or dask - they work just as before.
In cupy, we didn't do anything about the device argument in the end (#293 is DNM). We did tweak `to_device` though.
To the best of my knowledge pytorch now works 100%.

### Minor changes

- `finfo` and `iinfo` functions now accept array arguments, in accordance with the
   Array API spec;
- `torch.asarray` function propagates the device of the input array. This works around
   the [pytorch issue #150199](https://github.com/pytorch/pytorch/issues/150199);
- `torch.repeat` function is now available;
- `torch.count_nonzero` function now correctly handles the case of a tuple `axis`
  arguments and `keepdims=True`;
- `torch.meshgrid` wrapper defaults to `indexing="xy"`, in accordance with the
  array API specification;

> - `cupy.asarray` function now implements the `copy=True` argument;

Reviewer: ```suggestion
- `cupy.asarray` function now implements the `copy=False` argument, albeit
  at the cost of risking to make a temporary copy.
```

# Changelog

## 1.12.0 (2025-05-13)

### Major changes

- The build system has been updated to use `pyproject.toml` instead of `setup.py`
- Support for Python 3.9 has been dropped. The minimum supported Python version is now
  3.10; the minimum supported NumPy version is 1.22.
- The `linalg` extension works correctly with `pytorch>=2.7`.
- Multiple improvements to handling of devices in CuPy and PyTorch backends.
  Support for multiple devices in CuPy is still immature and you should use
  context managers rather than relying on input-output device propagation or

>   on the `device` parameter.

Author: ```suggestion
  on the `device` parameter.  Please report any issues you encounter.
```
